File: 410. Split Array Largest Sum/main.go
package main

import "fmt"

func main() {
	ans := splitArray([]int{1, 4, 4}, 3)
	fmt.Println(ans)
}

func splitArray(nums []int, m int) int {
	sum := 0
	max := -1
	for _, num := range nums {
		sum += num
		if num > max {
			max = num
		}
	}

	l, r := max, sum

	for l < r {
		mid := (l + r) / 2

		if minimumSubarraysRequired(nums, mid) <= m {
			r = mid
		} else {
			l = mid + 1
		}
	}

	return l
}

// minimumSubarraysRequired
func minimumSubarraysRequired(arr []int, x int) int {
	currSum := 0
	splitRequired := 0

	for _, num := range arr {
		if currSum+num <= x {
			currSum += num
		} else {
			currSum = num
			splitRequired++
		}
	}

	return splitRequired + 1
}

File: 410. Split Array Largest Sum/main.py
from typing import List


class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
        def split(max):
            count = 1
            sum = 0
            for n in nums:
                if sum + n > max:
                    count += 1
                    sum = n
                else:
                    sum += n
            return count

        def getMax():
            m = -1
            for n in nums:
                m = max(m, n)
            return m

        def getSum():
            m = 0
            for n in nums:
                m += n
            return m

        lo, hi = getMax(), getSum()
        while lo < hi:
            mid = lo + (hi - lo) // 2
            n = split(mid)

            if n == k:
                hi = mid
            elif n <= k:
                hi = mid
            else:
                lo = mid + 1
        return lo